src/wayland_keybinds.cpp

```diff
@@ -1,138 +1,313 @@
 
-static void seat_handle_capabilities(void *data, wl_seat *seat, uint32_t caps)
+    if (keymap) {
+        _xkb_state = xkb_state_new(keymap);
+        xkb_keymap_unref(keymap);
+    }
+}
+
+void SeatInfo::key(uint32_t key, uint32_t state)
 {
-   if(caps & WL_SEAT_CAPABILITY_KEYBOARD)
-   {
-      if(!keyboard)
-      {
-         keyboard = wl_seat_get_keyboard(seat);
-         wl_keyboard_add_listener(keyboard, &keyboard_listener, NULL);
-      }
-   }
+    if (!_xkb_state)
+        return;
+
+    xkb_keycode_t code = key + 8;
+    xkb_keysym_t sym = xkb_state_key_get_one_sym(_xkb_state, code);
+    if (sym == XKB_KEY_NoSymbol)
+        return;
+
+    if (state == WL_POINTER_BUTTON_STATE_PRESSED) {
+        _pressed_syms.push_back(sym);
+    } else if (state == WL_KEYBOARD_KEY_STATE_RELEASED) {
+        auto it = std::find(
+                _pressed_syms.begin(),
+                _pressed_syms.end(),
+                sym);
+        if (it != _pressed_syms.end())
+            _pressed_syms.erase(it);
+    }
+}
+
+void SeatInfo::modifiers(uint32_t depressed, uint32_t latched, uint32_t locked, uint32_t group)
+{
+    if (!_xkb_state)
+        return;
+
+    xkb_state_update_mask(_xkb_state, depressed, latched, locked, 0, 0, group);
+}
+
+void SeatInfo::clear()
+{
+    _pressed_syms.clear();
 }
 
 void update_wl_queue()
 {
-   wl_display_dispatch_queue_pending(wl_display_ptr, queue);
+    wl_display_dispatch_queue_pending(wl_display_ptr, queue);
 }
 
 void init_wayland_data()
 {
-   if (!wl_display_ptr)
-      return;
-
-   queue = wl_display_create_queue(wl_display_ptr);
-   struct wl_display *display_wrapped = (struct wl_display*)wl_proxy_create_wrapper(wl_display_ptr);
-   wl_proxy_set_queue((struct wl_proxy*)display_wrapped, queue);
-   wl_registry *registry = wl_display_get_registry(display_wrapped);
-   wl_proxy_wrapper_destroy(display_wrapped);
-   wl_registry_add_listener(registry, &registry_listener, NULL);
-   wl_display_roundtrip_queue(wl_display_ptr, queue);
-   wl_display_roundtrip_queue(wl_display_ptr, queue);
+    if (!wl_display_ptr)
+        return;
+
+    xkb_context = xkb_context_new(XKB_CONTEXT_NO_FLAGS);
+    if (!xkb_context)
+        return;
+
+    auto display_wrapped = static_cast<wl_display *>(wl_proxy_create_wrapper(wl_display_ptr));
+    if (!display_wrapped) {
+        xkb_context_unref(xkb_context);
+    }
+
+    queue = wl_display_create_queue(wl_display_ptr);
+    if (!queue) {
+        wl_proxy_wrapper_destroy(display_wrapped);
+        xkb_context_unref(xkb_context);
+        return;
+    }
+
+    wl_proxy_set_queue(reinterpret_cast<wl_proxy *>(display_wrapped), queue);
+
+    wl_registry_ptr = wl_display_get_registry(display_wrapped);
+    wl_proxy_wrapper_destroy(display_wrapped);
+
+    if (!wl_registry_ptr) {
+        wl_event_queue_destroy(queue);
+        xkb_context_unref(xkb_context);
+        return;
+    }
+
+    wl_registry_add_listener(wl_registry_ptr, &registry_listener, nullptr);
```

Comment:
wl_display_roundtrip_queue was here twice so that the seat is added before the function returns, should still work without it in this case but it would be simpler if everything was already setup when this function returned

Reply:
A roundtrip does not guarantee that, the only thing it guarantees is that any events the wl_display has received thus far are read and dispatched. The exception is the registry which the spec explicitly notes. Even then, not every seat will be in the registry when mangohud starts. 

I think it's fine to omit the round-trips.